metricbeat/scripts/mage/package.go

```diff
@@ -0,0 +1,232 @@
+// build/package/modules.d directories for use in packaging.
+func PrepareModulePackagingOSS() error {
+	return prepareModulePackaging([]struct{ Src, Dst string }{
+		{mage.OSSBeatDir("modules.d"), dirModulesDGenerated},
+	}...)
+}
+
+// PrepareModulePackagingXPack generates build/package/modules and
+// build/package/modules.d directories for use in packaging.
+func PrepareModulePackagingXPack() error {
+	return prepareModulePackaging([]struct{ Src, Dst string }{
+		{mage.OSSBeatDir("modules.d"), dirModulesDGenerated},
+		{"modules.d", dirModulesDGenerated},
+	}...)
+}
+
+func prepareModulePackaging(files ...struct{ Src, Dst string }) error {
+	mg.Deps(GenerateDirModulesD)
+
+	err := mage.Clean([]string{
+		dirModulesDGenerated,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, copyAction := range files {
+		err := (&mage.CopyTask{
+			Source:  copyAction.Src,
+			Dest:    copyAction.Dst,
+			Mode:    0644,
+			DirMode: 0755,
+		}).Execute()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+const modulesDHeader = "# Module: {{ .Module }}\n# Docs: https://www.elastic.co/guide/en/beats/{{ .BeatName }}/{{ .BeatDocBranch }}/{{ .BeatName }}-module-{{ .Module }}.html"
+
+// GenerateDirModulesD generates the modules.d directory
+func GenerateDirModulesD() error {
+	if err := os.RemoveAll("modules.d"); err != nil {
+		return err
+	}
+
+	shortConfigs, err := filepath.Glob("module/*/_meta/config.yml")
+	if err != nil {
+		return err
+	}
+	flavorConfigs, err := filepath.Glob("module/*/_meta/config-*.yml")
+	if err != nil {
+		return err
+	}
+	shortConfigs = append(shortConfigs, flavorConfigs...)
+
+	docBranch, err := mage.BeatDocBranch()
+	if err != nil {
+		errors.Wrap(err, "failed to get doc branch")
+	}
+
+	mode := 0644
+	for _, f := range shortConfigs {
+		moduleName, configName, ok := moduleConfigParts(f)
+		if !ok {
+			continue
+		}
+
+		suffix := ".yml.disabled"
+		if configName == "system" {
+			suffix = ".yml"
+		}
+		path := filepath.Join("modules.d", configName+suffix)
+
+		headerArgs := map[string]interface{}{
+			"Module":        moduleName,
+			"BeatName":      mage.BeatName,
+			"BeatDocBranch": docBranch,
+		}
+		header := mage.MustExpand(modulesDHeader, headerArgs)
+
+		err := copyWithHeader(header, f, path, os.FileMode(mode))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// moduleConfigParts obtain the moduleName and the configName from a config path.
+// The configName includes the flavor
+func moduleConfigParts(f string) (moduleName string, configName string, ok bool) {
+	parts := strings.Split(filepath.ToSlash(f), "/")
+	if len(parts) < 4 {
+		return
+	}
+	moduleName = parts[1]
+	configName = moduleName
+	ok = true
+
+	fileName := strings.TrimSuffix(parts[3], ".yml")
+	parts = strings.SplitN(fileName, "-", 2)
+	if len(parts) > 1 {
+		configName = moduleName + "-" + parts[1] // module + flavor
+	}
+	return
+}
+
+func copyWithHeader(header, src, dst string, mode os.FileMode) error {
```

Comment:
a comment here would probably help to understand what header is

Reply:
Sure :+1: